File: api/Services/GameData.cs
using System.Text.Json;
using api.Models;

namespace api.Services
{
    public class GameData
    {
        private const string DataDirectory = "data";

        private const string leaderboardFile = $"{DataDirectory}/leaderboard.json";

        /// <summary>
        /// Saves the game to a file.
        /// </summary>
        /// <param name="game">The game to save.</param>
        /// <param name="handle">The handle of the game.</param>
        public static async Task SaveGameAsync(Game game, string handle)
        {
            string filePath = $"{DataDirectory}/{handle}.json";
            var jsonString = JsonSerializer.Serialize(game);
            await File.WriteAllTextAsync(filePath, jsonString);
        }

        /// <summary>
        /// Loads the games from a file.
        /// </summary>
        /// <param name="filePath">The file path to load the games from.</param>
        /// <returns>The loaded games.</returns>
        public static async Task<Game?> RetrieveGameAsync(string handle)
        {
            string filePath = $"{DataDirectory}/{handle}.json";
            if (!File.Exists(filePath)) return null;
            var jsonString = await File.ReadAllTextAsync(filePath);
            return JsonSerializer.Deserialize<Game>(jsonString);
        }

        /// <summary>
        /// Saves the leaderboard entries to a file.
        /// </summary>
        /// <param name="leaderboardEntries">The leaderboard entries to save.</param>
        public static async Task SaveLeaderboardEntryAsync(List<Leaderboard> leaderboardEntries)
        {
            var jsonString = JsonSerializer.Serialize(leaderboardEntries);
            await File.WriteAllTextAsync(leaderboardFile, jsonString);
        }
        /// <summary>
        /// Loads the leaderboard entries from a file.
        /// </summary>
        /// <returns>The loaded leaderboard entries.</returns>
        public static async Task<List<Leaderboard>> RetrieveLeaderboardAsync()
        {
            if (!File.Exists(leaderboardFile)) return new List<Leaderboard>();
            var jsonString = await File.ReadAllTextAsync(leaderboardFile);
            return JsonSerializer.Deserialize<List<Leaderboard>>(jsonString) ?? new List<Leaderboard>();
        }
    }
}

File: app-java/src/main/java/com/example/app_java/Services/GameData.java
package com.example.app_java.Services;

import com.example.app_java.Models.Game;
import com.example.app_java.Models.Leaderboard;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

public class GameData {

    private static final String DATA_DIRECTORY = "data";
    private static final ObjectMapper objectMapper = new ObjectMapper();

    public static CompletableFuture<Void> saveGameAsync(Game game, String handle) {
        return CompletableFuture.runAsync(() -> {
            try {
                String sanitizedHandle = sanitizeHandle(handle);
                String filePath = Paths.get(DATA_DIRECTORY, sanitizedHandle + ".json").toString();
                if (!isPathWithinDirectory(filePath, DATA_DIRECTORY)) {
                    throw new IllegalStateException("Invalid file path.");
                }
                String jsonString = objectMapper.writeValueAsString(game);
                Files.write(Paths.get(filePath), jsonString.getBytes());
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    public static CompletableFuture<Game> retrieveGameAsync(String handle) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String sanitizedHandle = sanitizeHandle(handle);
                String filePath = Paths.get(DATA_DIRECTORY, sanitizedHandle + ".json").toString();
                if (!Files.exists(Paths.get(filePath)) || !isPathWithinDirectory(filePath, DATA_DIRECTORY)) {
                    return null;
                }
                String jsonString = new String(Files.readAllBytes(Paths.get(filePath)));
                return objectMapper.readValue(jsonString, Game.class);
            } catch (IOException e) {
                e.printStackTrace();
                return null;
            }
        });
    }

    public static CompletableFuture<Void> saveLeaderboardEntryAsync(String handle, int score) {
        return CompletableFuture.runAsync(() -> {
            try {
                String leaderboardFile = Paths.get(DATA_DIRECTORY, "leaderboard.json").toString();
                List<Leaderboard> leaderboard;
                if (new File(leaderboardFile).exists()) {
                    String leaderboardContent = new String(Files.readAllBytes(Paths.get(leaderboardFile)));
                    leaderboard = objectMapper.readValue(leaderboardContent, new TypeReference<List<Leaderboard>>() {
                    });
                } else {
                    leaderboard = new ArrayList<>();
                }
                leaderboard.removeIf(e -> e.getHandle().equals(handle));
                leaderboard.add(new Leaderboard(handle, score, LocalDateTime.now()));
                String updatedContent = objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(leaderboard);
                Files.write(Paths.get(leaderboardFile), updatedContent.getBytes());
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    public static CompletableFuture<List<Leaderboard>> retrieveLeaderboardAsync() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String leaderboardFile = Paths.get(DATA_DIRECTORY, "leaderboard.json").toString();
                if (!new File(leaderboardFile).exists()) {
                    return new ArrayList<>();
                }
                String jsonString = new String(Files.readAllBytes(Paths.get(leaderboardFile)));
                return objectMapper.readValue(jsonString, new TypeReference<List<Leaderboard>>() {
                });
            } catch (IOException e) {
                e.printStackTrace();
                return new ArrayList<>();
            }
        });
    }

    private static String sanitizeHandle(String handle) {
        return handle.replaceAll("\\.\\./", "").replaceAll("\\.\\\\", "");
    }

    private static boolean isPathWithinDirectory(String filePath, String directory) {
        try {
            String fullPath = new File(filePath).getCanonicalPath();
            String directoryPath = new File(directory).getCanonicalPath();
            return fullPath.startsWith(directoryPath);
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
    }
}